## Alice Balard, Eizaguirre lab, QMUL
## 22nd of June 2022
## Code structure for the project "Methylation inheritance after parasite infection in the three-spined stickleback"

############## DATA ORGANISATION  ################
### Raw files fro the 114 samples (zipped) here:
/data/archive/archive-SBCS-EizaguirreLab/RRBS/StickPara_Broject/00Illumina_RawReads/

[btx915@frontend8:/data/SBCS-EizaguirreLab/Alice/StickParaBroOff/]
├── Data
│   ├── 01TrimmedReads_cutadapt
│   ├── 02RefGenome
│   ├── 03Bismark_alignment
│   ├── 04BSBolt_methCall
│   ├── 05MethylKit
│   └── 06GynoGen_functionalAnnotation
├── GIT_StickParaOffsBroject
│   ├── code
│   ├── data
│   ├── gitignore
│   └── ReadMe.Rmd
└── QualityControl
    ├── QC00_RawReads_fastqc
    └── QC01_TrimmedReads_fastqc

########## BASH SCRIPTS (within GIT_StickParaOffsBroject/code) ################
├── miniBashTools
├── misc
├── R
├── run_gynofuncal.stderr
├── run_gynofuncal.stdout
├── S00_multiQC_generateReports.sh
├── S01_fastqc.sh
├── S02.1_adapterTrim_cutadapt.sh
├── S02.2_fastqcTrimmed.sh
├── S03.1_BSBolt_index.sh
├── S03.2_BSBolt_align.sh
├── S03.3_BSBolt_SortIndexBam.sh
├── S03.4_BSBolt_MethCall_forloop.sh
├── S03.5_prepareBSBoltOutput.sh # prepare BSBolt for further R steps, cut by CG, CHH etc.
└── S04_GynoGen_functionalAnnotation.sh # add blastp gene names and gene ontology to Miles & Kostas gynogen genome

########## R SCRIPTS (within GIT_StickParaOffsBroject/code/R) ################

├── customRfunctions.R
├── librariesLoading.R
├── R00_calculateStats.R # creat fullmetadata135 and compare BSBolt with Bismark mapping
├── R01.1_prepBSBOLTForMethylkit.R
├── R01.2_prepObjectMethylkit.R
├── R02.1_loadMetadata.R
├── R02.2_loadMethyldata.R
├── R02.3_DATALOAD.R # main loading script, call R02.1 and R02.2
├── R03.1_methylationAndFitness.Rmd # Explore link between global methylation and body condition index
├── R03.2_globalMethylation.R
├── R04.1.2_getDifferentialMethylationByBrotherPairs.R
├── R04.1_getDifferentialMethylation.R
├── R04.2_exploreDifferentialMethylation.R
├── R04.3_GeneOntology.R
├── R04.4_NetworkAnalyses.R
├── rawDataCleaning.R
├── Rdata
├── readme.txt
├── Rfigures
├── run_R00.R.stdout
├── run_R00.sh
├── run_R01.2.R.sh
├── run_R01.2.R.stdout
├── run_R03.1.R_oldglobalmethylation.sh
├── run_R03.1.R.stdout
├── run_R04.1.2_getDifferentialMethylationByBrotherPairs.R
├── run_R04.1.2_getDifferentialMethylationByBrotherPairs.R.o2434786
├── run_R04.1.diffMeth.R.sh
├── run_R04.1.diffMeth.R.stdout
└── Rxx_sliddingWindow.R




##############################
##### R methylation analysis workflow: #####
Preparation of data and metadata:
R00 -> alignment stats
R01.1, R01.2, R01.3, R01.4 -> data preparation

R02 -> link methylation and fitness (tbc)

# PART 1. (script R03.1) Methylation profile: or Adonis multivariate stats, presence/abs, block effect=family
# PART 2. (script R03.2) DMS 
# PART 3.DMR: Precise localisation of blocks investing in methylation
# PART 4.Network: Explore other parameters than geographic: are the modules
#   on a high recombination place? Or with high mutation rate? 
#   What does modules capture on top of DMRs?
##### Then 
# - Gene Ontology
# - Final: Check robustness: reshuffle the controls and check that we find less DMS/DMR/modules
# 
# These analyses will be done on:
# A. Parents ctr-trt
# B. Offspring ctr-trt pat1 & ctrl-trt pat2 (different offspring treatments)
# C. Offspring ctrl-ctrl & trt-trt (different paternal treatment)
##############################
